feat: add theta cache for Rope

mzbac opened the pull request:
## Proposed changes

this PR is addressing https://github.com/ml-explore/mlx/issues/360

**Test**:
Manually tested again Llama-7B (scale 1.0) and Deepseek Coder 6.7B (scale 1/4.0), both work.

**Note**:
I have run the llama_mlx benchmark test and haven't found any performance degradation. The difference between cached vs no cache is only around 0.01-0.02ms per token.

## Checklist

Put an `x` in the boxes that apply.

- [x] I have read the [CONTRIBUTING](https://github.com/ml-explore/mlx/blob/main/CONTRIBUTING.md) document
- [x] I have run `pre-commit run --all-files` to format my code / installed pre-commit prior to committing changes
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have updated the necessary documentation (if needed)

mzbac commented on python/mlx/nn/layers/positional_encoding.py:
```diff
@@ -25,20 +25,35 @@ class RoPE(Module):
         traditional: bool = False,
         base: float = 10000,
         scale: float = 1.0,
+        max_position: int = 2048,
```
To enable building the theta during initialization, we need to know the maximum position and data type.

mzbac commented on python/mlx/nn/layers/positional_encoding.py:
```diff
@@ -75,31 +90,43 @@ def __call__(self, x, offset: int = 0):
-        costheta, sintheta = RoPE.create_cos_sin_theta(
-            N, self.dims, offset=offset, base=self.base, scale=self.scale, dtype=x.dtype
-        )
+        if N > self.max_position:
```
I am not sure if we need to cover the scenario where the sequence length is greater than the maximum position here. This could only happen when user passes the wrong maximum position during initialization. Since we are in the relatively early stage of mlx framework, maybe we should just throw an error here when the sequence length exceeds the maximum position so that people can fix their code.

mzbac commented on python/mlx/nn/layers/positional_encoding.py:
```diff
@@ -75,31 +90,43 @@ def __call__(self, x, offset: int = 0):
         return mx.reshape(rx, shape)
 
-    @staticmethod
-    def create_cos_sin_theta(
```
rewrite the `create_cos_sin_theta` function because we only want to precompute the theta once.

mzbac commented on python/mlx/nn/layers/positional_encoding.py:
```diff
@@ -75,31 +90,43 @@ def __call__(self, x, offset: int = 0):
-    def create_cos_sin_theta(
-        N: int,
-        D: int,
-        offset: int = 0,
```
Remove the offset because it will keep changing during inference.

angeloskath commented:
Hm I think I don't like having a maximum sequence length in RoPE or slicing in the costheta and sintheta. Basically amortized over the number of layers it doesn't matter that costheta and sintheta are going to be computed once per token. They are not a big performance bottleneck anyway, the actual computation of the rotations takes way more time.

So my suggestion would be to simply cache the output of `create_cos_sin_theta` based on its arguments rather than replace it with something more complicated. Similarly, you can cache the output of `create_alibi_matrix` and add both on this PR.

mzbac commented:
@angeloskath Thanks for the feedback. This may explain why I can't see much improvement in the benchmark test. I will try your suggestion.

mzbac commented:
Hi @angeloskath I have updated the cache based on your suggestion. I didn't notice any performance degradation (nor much improvement either). also added a cache for create_alibi_matrix, but I wouldn't test it as I can't find its use case example, but I assume the simple change should work.

angeloskath commented:
Nice it looks good. I will review it shortly. If it had a big impact we'd have added it but in general it should be strictly better so no reason not to add it. At the very least it reduces the size of the graph we have to process as the `cos(θ)` and `sin(θ)` are reused in every layer.

mzbac commented:
Yeah, the cache should definitely improve performance. The reason I didn't see much improvement on the benchmark may be as you said, it's not the bottleneck for the llama model inference. Thanks for helping to review the PR :)
